chain/index/read.go

```diff
@@ -0,0 +1,135 @@
+package index
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+const headIndexedWaitTimeout = 5 * time.Second
+
+func (si *SqliteIndexer) GetCidFromHash(ctx context.Context, txHash ethtypes.EthHash) (cid.Cid, error) {
+	if si.isClosed() {
+		return cid.Undef, ErrClosed
+	}
+
+	var msgCidBytes []byte
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
+		return si.queryMsgCidFromEthHash(ctx, txHash, &msgCidBytes)
+	}); err != nil {
+		return cid.Undef, err
+	}
+
+	msgCid, err := cid.Cast(msgCidBytes)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to cast message CID: %w", err)
+	}
+
+	return msgCid, nil
+}
+
+func (si *SqliteIndexer) queryMsgCidFromEthHash(ctx context.Context, txHash ethtypes.EthHash, msgCidBytes *[]byte) error {
+	return si.stmts.getMsgCidFromEthHashStmt.QueryRowContext(ctx, txHash.String()).Scan(msgCidBytes)
+}
+
+func (si *SqliteIndexer) GetMsgInfo(ctx context.Context, messageCid cid.Cid) (*MsgInfo, error) {
+	if si.isClosed() {
+		return nil, ErrClosed
+	}
+
+	var tipsetKeyCidBytes []byte
+	var height int64
+
+	if err := si.readWithHeadIndexWait(ctx, func() error {
```

Comment:
`readWithHeadIndexWait` might be overkill for this function, the downstream consumer of this function, `StateManager#SearchForMessage` and `StateManager#WaitForMessage` have fallbacks if we can't find it, and their fallbacks may end up being quicker than waiting because they start searching at the head tipset for the message--so if we're waiting for the head tipset to be properly indexed before we can return, it may take longer than the caller just looking in the head tipset to find it.

Reply:
Done.